data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1262 @@
+            // a rolling "compander" mechanism. That is, we keep track of the absolute range of
+            // lowest to highest priority values for all the previous epochs, to then compare the
+            // range in each epoch and map one range to another to find its relative position.
+            let drt_lowest_compared_to_absolute = drt_lowest / drt_lowest_absolute;
+            let drt_range_absolute = drt_highest_absolute / drt_lowest_absolute;
+            let vtt_lowest_compared_to_absolute = vtt_lowest / vtt_lowest_absolute;
+            let vtt_range_absolute = vtt_highest_absolute / vtt_lowest_absolute;
+            let drt_bucket_index =
+                drt_lowest_compared_to_absolute / drt_range_absolute * buckets_count;
+            let vtt_bucket_index =
+                vtt_lowest_compared_to_absolute / vtt_range_absolute * buckets_count;
+
+            // For a perfect calculation, all values lower than the lowest bucket index
+            // (representing the lowest fee should be inserted. However, we can get a good enough
+            // approximation while saving CPU and memory by inserting only the 10% closest values
+            // from below.
+            for bucket_index in (drt_bucket_index * 0.9) as u64..=drt_bucket_index as u64 {
+                drt_counter.add(bucket_index);
+            }
+            for bucket_index in (vtt_bucket_index * 0.9) as u64..=vtt_bucket_index as u64 {
+                vtt_counter.add(bucket_index);
+            }
+        }
+
+        // Make an estimation for each of the targeted time-to-blocks.
+        let mut drt_priorities: Vec<Priority> = vec![];
+        let mut vtt_priorities: Vec<Priority> = vec![];
+        for minutes in target_minutes.into_iter() {
+            // Derive the frequency threshold for this targeted time-to-block.
+            let epochs = f64::from(minutes) * 60.0 / f64::from(seconds_per_epoch);
+            let epochs_freq = epochs / priorities_count;
+            let threshold = epochs_freq / bucket_capacity;
+
+            // Run the frequency query on the lossy counters.
+            let drt_elements = drt_counter.query(threshold);
+            let vtt_elements = vtt_counter.query(threshold);
+
+            // The priority is calculated by reverting the buckets mapping performed before, i.e.
+            // mapping the bucket index back to a priority value.
+            let drt_bucket = drt_elements.max().unwrap_or_default();
+            let drt_priority = drt_lowest_absolute.unwrap_or_default()
+                + (drt_highest_absolute - drt_lowest_absolute.unwrap_or_default())
+                    * (drt_bucket as f64 / buckets_count);
+            let vtt_bucket = vtt_elements.max().unwrap_or_default();
+            let vtt_priority = vtt_lowest_absolute.unwrap_or_default()
+                + (vtt_highest_absolute - drt_lowest_absolute.unwrap_or_default())
+                    * (vtt_bucket as f64 / buckets_count);
```

Comment:
There is a bug here:

```suggestion
            let vtt_priority = vtt_lowest_absolute.unwrap_or_default()
                + (vtt_highest_absolute - vtt_lowest_absolute.unwrap_or_default())
                    * (vtt_bucket as f64 / buckets_count);
```